doc/ipv6/ipv6_link_local.md

```diff
@@ -0,0 +1,378 @@
+
+   `#config interface ipv6 enable use-link-local-only Ethernet24`
+
+   `#config interface ipv6 disable use-link-local-only Ethernet24`
+
+4. Provide SONiC CLI global command to enable or disable the IPv6 auto link-local address generation configuration on all eligible interfaces.
+
+   `#config ipv6 enable`
+
+   `#config ipv6 disable`
+
+5. Standard SNMP IP-MIB should support retrieving of link-local IPv6 addresses/neighbors.
+
+
+### 1.1.3 Scalability Requirements
+The maximum number of IPv6 link-local addresses depends on number of Ethernet, LAG, VLAN interfaces supported by the switch. For example, a 64 port switch with 64 max PortChannel interfaces and 4K VLAN interfaces would require maximum of 4094+64+64 link-local addresses.
+
+### 1.1.4 Warm Boot Requirements
+L3 Traffic destined to IPv6 link-local next hops/neighbors should continue to be routed across warm restart. There should be no or minimal loss with in acceptable limits (< 1sec).
+
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+SONiC already supports IPv6 addressing. This feature extends IPv6 by adding link-local address support.  SWSS is modified to handle IPv6 link-local neighbors and IPv6 routes with link-local next hops.
+
+### 1.2.2 Container
+The code changes for this feature are part of *neighsyncd*  *routeorch* and *intfmgrd* modules which are part of SWSS container.
+
+### 1.2.3 SAI Overview
+There are no changes to SAI headers/implementation to support this feature. Existing SAI Neighbor, Next hop and Route APIs are used to handle IPv6 link-local addresses.
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+When BGP is deployed in a IPv6 data center network, it is desirable to configure BGP neighbors using IPv6 link-local addresses to avoid assigning and managing global IPv6 addresses on interconnect links.  In such configurations, BGP auto-detects the neighbor using IPv6 router advertisements. This can be leveraged to advertise IPv4 prefixes over IPv6 BGP session (RFC5549).
+
+The below diagram shows the deployment use case.
+
+1. IPv6 link local address used for BGP peering
+2. RFC 5549 advertises the IPv4 addresses over BGP IPv6 peering
+3. IPv6 router advertisement is used to learn peer's link-local address.
+4. For configuration, the BGP neighbor is identified using interface 'Ethernet0' and not IPv6 address.
+
+![Deployment use case](images/bgp.jpg)
+
+The following command shows the IPv4 routes learned via IPv6 link-local next hops on Router A:
+
+```
+sonic# show ip route
+Codes: K - kernel route, C - connected, S - static, R - RIP,
+       O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
+       T - Table, v - VNC, V - VNC-Direct, A - Babel, D - SHARP,
+       F - PBR, f - OpenFabric,
+       > - selected route, * - FIB route # - not installed in hardware
+
+B    192.168.0.0/24 [20/0] via fe80::5054:ff:fe03:6175, Ethernet0, 00:00:06

```

Comment:
Since FRR/zebra is already pushing 169.254.0.1 IPv4 link local NH for IPv4 route to kernel, let's use the same for HW programming flow (fpmsyncd) as well, we're already listening neighbor events from kernel, the existing code should work for this case as well, please share the issues you have encountered.

Reply:
The unnumbered routes needs to be sent to fpm with special attribute (e.g. onlink) to restrict next-hop resolution to be confined to the link on which route is getting programmed. Zebra doesn't pass any such attribute, and route orch doesn't have such ability to handle that today. Also, same neighbor entry 169.254.0.1 will appear on multiple interfaces, and this is also something not expected and requires special handling in sonic. Note that not supporting rfc 5549 next-hops natively is a limitation in Linux kernel and what zebra does is a hack.

Keeping all of the above into consideration, zebra is patched to send ipv6 next-hop along with the route. This approach is much cleaner and avoids any special neigh/next-hop handling in neigh/route orch.

Reply:
Hasan - Agree that this solution will work, trying to understand the advantage of deviating the kernel & FPM flow behaviors, I'm not aware of any FRR user other than SONiC already deviating like this by introducing a change in the FRR module, I believe the change in FRR will stay as PATCH and we wont be able to merge with FRR opensource code i.e somone in SONiC has to maintain the SONiC specific FRR changes (Any discussion with FRR team already happened on this change?)

>> The unnumbered routes needs to be sent to fpm with special attribute (e.g. onlink) to restrict next-hop resolution to be confined to the link on which route is getting programmed.

How's this handled in kernel, I believe we dont expect the NH resolution in kernel either, if FRR restricts it by adding the static Ipv4 link-local neighbor(169.254.0.1), cant we follow the same in FPM flow as well.

Also, same neighbor entry 169.254.0.1 will appear on multiple interfaces, and this is also something not expected and requires special handling in sonic.

>> This is not different from any IPv6 LLA neighbor i.e same LLA can be present on multiple interfaces,  if we use "SAI_NEIGHBOR_ENTRY_ATTR_NO_HOST_ROUTE" already for IPv6 LLA nbr, why cant we use the same for IPv4 LLA neighbor?

Reply:
Can we add this details to the HLD section ?